lib/worker.ex

```diff
@@ -3,43 +3,60 @@ defmodule Monarch.Worker do
   Defines the Monarch Worker for running Oban jobs.
   """
 
-  use Oban.Worker
+  use Oban.Worker,
+    unique: [
+      fields: [:args, :worker],
+      keys: [:job, :repo]
+    ]
 
+  @impl Oban.Worker
   def perform(job) do
     worker = String.to_existing_atom(job.args["job"])
     repo = String.to_existing_atom(job.args["repo"])
 
     # Wraps a chunk of records to be updated in a transaction
     {:ok, action} =
-      if apply(worker, :skip, []) do
-        {:ok, :halt}
-      else
-        repo.transaction(fn ->
-          case apply(worker, :query, []) do
-            [] ->
-              :halt
-
-            chunk ->
-              apply(worker, :update, [chunk])
-              :cont
-          end
-        end)
+      cond do
+        apply(worker, :skip, []) ->
+          {:ok, :halt}
+
+        Monarch.completed?(repo, worker) ->
+          {:ok, :halt}
+
+        true ->
+          repo.transaction(fn ->
+            case apply(worker, :query, []) do
+              [] ->
+                :ok = log_completed!(repo, worker)
```

Comment:
We historically didn't do this in the transaction so we might have had race conditions here